### PR TITLE
Bug 1822396: Delete subscription metric when an operator is uninstalled

### PR DESCRIPTION
**Pull request by anik120**
When an operator was subscribed to using a Subscription Object, the subscription_sync_total
metric was emitted whenever the Subscription Object was created/updated/deleted. This PR
updates that behaviour to emit the metric only when the Subscription object is created/updated,
and deletes the time series for that particular subscription when the subscription object is
deleted.

**Description of the change:**

**Motivation for the change:**

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: pkg/metrics/metrics.go [hunk 1]
```diff
@@ -212,3 +208,16 @@ func EmitCSVMetric(oldCSV *olmv1alpha1.ClusterServiceVersion, newCSV *olmv1alpha
 		csvAbnormal.WithLabelValues(newCSV.Namespace, newCSV.Name, newCSV.Spec.Version.String(), string(newCSV.Status.Phase), string(newCSV.Status.Reason)).Set(1)
 	}
 }
+
+func EmitSubsMetrics(sub *olmv1alpha1.Subscription) {
+	// sub.Spec is not a required field.
+	if sub.Spec == nil {
+		return
+	}
```
Comment (awgreene) [inline pkg/metrics/metrics.go:216]: As far as I can tell, this is no longer true. I tried applying the following files:
```
$ cat empty.sub.cr.yaml 
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: my-operator-subsription
  namespace: default

$ oc apply -f empty.sub.cr.yaml 
The Subscription "my-operator-subsription" is invalid: spec: Required value
```

&

```
$ cat empty.sub.cr.yaml 
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: my-operator-subsription
  namespace: default
spec: {}

$ oc apply -f empty.sub.cr.yaml 
The Subscription "my-operator-subsription" is invalid: 
* spec.name: Required value
* spec.source: Required value
* spec.sourceNamespace: Required value
```

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -212,3 +208,16 @@ func EmitCSVMetric(oldCSV *olmv1alpha1.ClusterServiceVersion, newCSV *olmv1alpha
...
+
+	SubscriptionSyncCount.WithLabelValues(sub.GetName(), sub.Status.InstalledCSV, sub.Spec.Channel, sub.Spec.Package).Inc()
+}
+
+func DeleteSubsMetric(sub *olmv1alpha1.Subscription) {
+	SubscriptionSyncCount.DeleteLabelValues(sub.GetName(), sub.Status.InstalledCSV, sub.Spec.Channel, sub.Spec.Package)
+}
```
Comment (awgreene) [inline pkg/metrics/metrics.go:223]: Can we add a test to the [metrics-e2e test](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/test/e2e/metrics_e2e_test.go)? The test could:
- Create a sub
- check for the metric
- update the sub
- check for the metric
- delete the sub
- ensure the metrics does not exist

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -212,3 +208,16 @@ func EmitCSVMetric(oldCSV *olmv1alpha1.ClusterServiceVersion, newCSV *olmv1alpha
...
 	}
 }
+
+func EmitSubsMetrics(sub *olmv1alpha1.Subscription) {
+	// sub.Spec is not a required field.
+	if sub.Spec == nil {
+		return
+	}
+
+	SubscriptionSyncCount.WithLabelValues(sub.GetName(), sub.Status.InstalledCSV, sub.Spec.Channel, sub.Spec.Package).Inc()
```
Comment (awgreene) [inline pkg/metrics/metrics.go:218]: Note: sub.Spec.Channel and sub.Spec.Package are not required fields - these could be empty strings. Should we emit these metrics?

File: pkg/metrics/metrics.go [hunk 4]
```diff
@@ -212,3 +208,16 @@ func EmitCSVMetric(oldCSV *olmv1alpha1.ClusterServiceVersion, newCSV *olmv1alpha
...
+}
+
+func DeleteSubsMetric(sub *olmv1alpha1.Subscription) {
+	SubscriptionSyncCount.DeleteLabelValues(sub.GetName(), sub.Status.InstalledCSV, sub.Spec.Channel, sub.Spec.Package)
```
Comment (awgreene) [inline pkg/metrics/metrics.go:222]: Note: sub.Spec.Channel and sub.Spec.Package are not required fields - these could be empty strings. Should we emit these metrics?

Comment (awgreene) [inline pkg/metrics/metrics.go:222]: sub.Spec.Channel could be an empty string, but we should be safe here because the CSV name is probably unique. Thoughts?

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (awgreene) [reply pkg/metrics/metrics.go:218]: @anik120 this is the only non-nit that I have on the PR, great work.

**Comment by anik120** [comment]
@awgreene thanks for the review! I have addressed the nit comments, and I'm waiting for https://github.com/operator-framework/operator-lifecycle-manager/pull/1443/ to merge before I write the test you suggested.

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (anik120) [reply pkg/metrics/metrics.go:213]: @awgreene what is the issue if we emit empty string?

File: pkg/metrics/metrics.go (hunk 2 above)
Comment (anik120) [reply pkg/metrics/metrics.go:228]: @awgreene added an e2e test case here: https://github.com/operator-framework/operator-lifecycle-manager/pull/1519/files#diff-41336190eb5305cb6fc38321ea468723R70-R84

**Comment by anik120** [comment]
>@awgreene thanks for the review! I have addressed the nit comments, and I'm waiting for #1443 to merge before I write the test you suggested.

Since #1443 is not merged yet, adding to the existing portfolio that'll need to be incorporated into #1443

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (awgreene) [reply pkg/metrics/metrics.go:223]: Should we waste telemeter bandwith reporting a sub that doesn't have:

- A name
- A CSV that it installs
- Define a package or channel to pull from?

File: test/e2e/metrics_e2e_test.go [hunk 5]
```diff
@@ -80,6 +97,18 @@ func getOLMPod(client operatorclient.ClientInterface) *corev1.Pod {
 	return &podList.Items[0]
 }
 
+func getCatalogPod(client operatorclient.ClientInterface) *corev1.Pod {
```
Comment (awgreene) [inline test/e2e/metrics_e2e_test.go:100]: This function seems to be extremely specific to a single test. It should probably be made more generic (takes a label) or just implemented inline.

Comment (anik120) [reply test/e2e/metrics_e2e_test.go]: @awgreene good catch. Updated the PR with a generic function.

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (anik120) [reply pkg/metrics/metrics.go:223]: @awgreene  isn't name, installedCSV required fields?

File: pkg/controller/operators/catalog/subscription/syncer.go [hunk 6]
```diff
@@ -45,34 +45,37 @@ func (s *subscriptionSyncer) now() *metav1.Time {
 // Sync reconciles Subscription events by invoking a sequence of reconcilers, passing the result of each
 // successful reconciliation as an argument to its successor.
 func (s *subscriptionSyncer) Sync(ctx context.Context, event kubestate.ResourceEvent) error {
-	res := &v1alpha1.Subscription{}
-	if err := scheme.Convert(event.Resource(), res, nil); err != nil {
+	sub := &v1alpha1.Subscription{}
```
Comment (ecordell) [inline pkg/controller/operators/catalog/subscription/syncer.go:48]: these changes are unrelated to the bug (and pretty subjective), can you remove them so that we can focus on the fix when we are backporting?

Comment (anik120) [reply pkg/controller/operators/catalog/subscription/syncer.go]: @ecordell removed the changes (although I'd have liked to discuss the pros and cons of making changes like these if we had a bit more time)

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (awgreene) [reply pkg/metrics/metrics.go:223]: >If the Subscription is being created on cluster successfully, I'm not convinced we should choose tweak the metrics to not emit data point based on field values.
That would essentially mean we have subscriptions on cluster, but telemeter is unaware of them.

If we can use those incomplete metrics we omit to find identify something useful, then yes, we should emit them. Otherwise, we're just consuming our telemeter quota for no reason.

Comment (anik120) [reply pkg/metrics/metrics.go:223]: @awgreene so in summary: 

a) We could either not emit those metrics now, and if we run into a use case where we need those data points we can start emitting them too. But then they'd have to be backported into previous z stream releases (which we already know is an EXTREMELY delayed process, compared to how rapidly telemeter dashboards can be created)

b) We could just emit the metric since a subscription is being created (and note that it'll be deleted now once the subscription is deleted) which will use up some of the telemeter quota. Question here is: how often are we expecting these kind of subscriptions to be created (my guess is they won't be created in any prod clusters)? And as a result, how much quota are we expecting to use up for this edge case? 

I just want to make sure we don't put our cart before our horses and potentially expose ourselves to a bug back porting process.

File: pkg/controller/operators/catalog/subscription/syncer.go [hunk 7]
```diff
@@ -69,7 +69,10 @@ func (s *subscriptionSyncer) Sync(ctx context.Context, event kubestate.ResourceE
 	case kubestate.ResourceUpdated:
 		initial = initial.Update()
 	case kubestate.ResourceDeleted:
-		initial = initial.Delete()
+		{
```
[the remark this answers is not in the document]
Comment (anik120) [reply pkg/controller/operators/catalog/subscription/syncer.go:72]: @awgreene yes. Keeping the statements for this `case` in a block of {}

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (awgreene) [reply pkg/metrics/metrics.go:223]: I am unconvinced that there will ever be a point that we will want to consume the metric when the installedCSV is blank.

However, given that the metrics are deleted when the subscription is deleted, I see no way that the metric could blow up. I also agree with your point about the ease of updating telemeter vs backporting changes. Given those points, it probably isn't worth delaying this bug fix.

File: pkg/controller/operators/catalog/subscription/syncer_test.go [hunk 8]
```diff
@@ -6,11 +6,19 @@ import (
...
 	"github.com/operator-framework/api/pkg/operators/v1alpha1"
 	"github.com/operator-framework/operator-lifecycle-manager/pkg/lib/kubestate"
 )
 
+const (
+	name             = "test-subscription"
+	packageName      = "test-package"
+	channel          = "test-channel"
+	installedCSVName = "test-csv"
+)
```
Comment (awgreene) [inline pkg/controller/operators/catalog/subscription/syncer_test.go:20]: non-blocking nit - these strings are only used in one place, could be inline.

File: pkg/controller/operators/catalog/subscription/syncer.go [hunk 9]
```diff
@@ -45,34 +45,37 @@ func (s *subscriptionSyncer) now() *metav1.Time {
...
 	}
 
-	s.recordMetrics(res)
+	metrics.EmitSubsMetrics(subs)
 
 	logger := s.logger.WithFields(logrus.Fields{
-		"reconciling": fmt.Sprintf("%T", res),
-		"selflink":    res.GetSelfLink(),
+		"reconciling": fmt.Sprintf("%T", subs),
+		"selflink":    subs.GetSelfLink(),
```
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/controller/operators/catalog/subscription/syncer.go]: FYI, selflink is deprecated now.

File: pkg/controller/operators/catalog/subscription/syncer.go [hunk 10]
```diff
@@ -70,6 +70,7 @@ func (s *subscriptionSyncer) Sync(ctx context.Context, event kubestate.ResourceE
 		initial = initial.Update()
 	case kubestate.ResourceDeleted:
 		initial = initial.Delete()
+		metrics.DeleteSubsMetric(res)
```
Comment (ecordell) [inline pkg/controller/operators/catalog/subscription/syncer.go:73]: This only deletes the time series when the subscription has been deleted, but from the bug report:

> Metrics for Subscriptions of CSVs that have been upgraded to new versions still hang around after new Subscription/CSV have succeeded. 

Because installedCSV is one of the labels/values, we would need to delete the metric (with any non-current installedCSV) on upgrade as well. Or have I misunderstood the bug?

Comment (anik120) [reply pkg/controller/operators/catalog/subscription/syncer.go:81]: @ecordell removing the metrics for the old subscription when a subscription was the original intention of the bug. 

Updated the PR.

File: test/e2e/metrics_e2e_test.go [hunk 11]
```diff
@@ -61,15 +69,46 @@ var _ = Describe("Metrics", func() {
...
-		Expect(getMetricsFromPod(c, getOLMPod(c), "8081")).ToNot(And(
+		Expect(getMetricsFromPod(c, getPodWithLabel(c, "app=olm-operator"), "8081")).ToNot(And(
 			ContainSubstring("csv_abnormal{name=\"%s\"", failingCSV.Name),
 			ContainSubstring("csv_succeeded{name=\"%s\"", failingCSV.Name),
 		))
+
+		subscriptionCleanup, subscription := createSubscriptionWithInstalledCSV(GinkgoT(), crc, testNamespace, "metric-subscription", testPackageName, stableChannel, installedCSV, v1alpha1.ApprovalManual)
+
+		// Verify metrics have been emitted for subscription
+		Expect(getMetricsFromPod(c, getPodWithLabel(c, "app=catalog-operator"), "8081")).To(And(
```
Comment (njhale) [inline test/e2e/metrics_e2e_test.go:80]: Should these checks use the `Eventually` matcher instead? If I'm reading this correctly, if the metric isn't emitted on time, then the test will fail.

File: pkg/controller/operators/catalog/subscription/syncer.go [hunk 12]
```diff
@@ -67,9 +69,21 @@ func (s *subscriptionSyncer) Sync(ctx context.Context, event kubestate.ResourceE
 	case kubestate.ResourceAdded:
 		initial = initial.Add()
 	case kubestate.ResourceUpdated:
+		oldSubscription, err := s.client.OperatorsV1alpha1().Subscriptions(res.GetNamespace()).Get(context.TODO(), res.GetName(), metav1.GetOptions{})
+		if err != nil {
+			logger.Errorf("Could not get old subscription %s in namespace %s. Skipping attempt to delete metrics", oldSubscription.GetName(), oldSubscription.GetNamespace())
+		} else {
+			if res.GetName() != oldSubscription.GetName() ||
```
Comment (njhale) [inline pkg/controller/operators/catalog/subscription/syncer.go:76]: `res.GetName() != oldSubscription.GetName()` should never happen, right?

Also, based on https://github.com/operator-framework/operator-lifecycle-manager/pull/1519/files#r431776633, I think we want to make sure `currentCSV != installedCSV` before deleting.

@ecordell does this line up with your comment?

File: pkg/controller/operators/catalog/subscription/syncer.go [hunk 13]
```diff
@@ -66,10 +66,19 @@ func (s *subscriptionSyncer) Sync(ctx context.Context, event kubestate.ResourceE
 	switch event.Type() {
 	case kubestate.ResourceAdded:
 		initial = initial.Add()
+		metrics.EmitSubMetric(res)
 	case kubestate.ResourceUpdated:
+		oldSubscription, err := s.client.OperatorsV1alpha1().Subscriptions(res.GetNamespace()).Get(context.TODO(), res.GetName(), metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (anik120) [reply pkg/controller/operators/catalog/subscription/syncer.go:72]: @njhale could you explain what you mean by "from the cache"? The only subscription I'm aware of is the one on cluster, which I decided to grab and delete the metrics for before updating the on cluster subscription.

File: pkg/metrics/metrics.go [hunk 14]
```diff
@@ -217,3 +218,11 @@ func EmitCSVMetric(oldCSV *olmv1alpha1.ClusterServiceVersion, newCSV *olmv1alpha
 		csvAbnormal.WithLabelValues(newCSV.Namespace, newCSV.Name, newCSV.Spec.Version.String(), string(newCSV.Status.Phase), string(newCSV.Status.Reason)).Set(1)
 	}
 }
+
+func EmitSubMetric(sub *olmv1alpha1.Subscription) {
+	SubscriptionSyncCount.WithLabelValues(sub.GetName(), sub.Status.InstalledCSV, sub.Spec.Channel, sub.Spec.Package).Inc()
```
[the remark this answers is not in the document]
Comment (anik120) [reply pkg/metrics/metrics.go:223]: @njhale I see your point. If we have consensus that this PR will be a good place to include the namespace label [here](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/metrics/metrics.go#L153) I don't mind adding it.

File: test/e2e/metrics_e2e_test.go [hunk 15]
```diff
@@ -61,15 +69,46 @@ var _ = Describe("Metrics", func() {
...
+			ContainSubstring("subscription_sync_total"),
+			ContainSubstring(fmt.Sprintf("%s=\"%s\"", metrics.NAME_LABEL, "metric-subscription")),
+			ContainSubstring(fmt.Sprintf("%s=\"%s\"", metrics.CHANNEL_LABEL, stableChannel)),
+			ContainSubstring(fmt.Sprintf("%s=\"%s\"", metrics.INSTALLED_LABEL, installedCSV)),
+			ContainSubstring(fmt.Sprintf("%s=\"%s\"", metrics.PACKAGE_LABEL, testPackageName)),
+		))
+
+		subscription.Status.InstalledCSV = installedCSVUpdated
```
[the remark this answers is not in the document]
Comment (anik120) [reply test/e2e/metrics_e2e_test.go]: @njhale I see what you mean https://github.com/operator-framework/operator-lifecycle-manager/pull/1519/checks?check_run_id=718197936#step:3:1582

Any ideas on how we could test the behavior on update in that case?

File: test/e2e/metrics_e2e_test.go [hunk 16]
```diff
@@ -117,3 +156,35 @@ func getMetricsFromPod(client operatorclient.ClientInterface, pod *corev1.Pod, p
...
+		},
+	}
+
+	subscription, err := crc.OperatorsV1alpha1().Subscriptions(namespace).Create(context.TODO(), subscription, metav1.CreateOptions{})
+	require.NoError(t, err)
+	return buildSubscriptionCleanupFunc(crc, subscription), subscription
+}
+
+func updateSubscription(t GinkgoTInterface, crc versioned.Interface, subscription *v1alpha1.Subscription) {
+	_, err := crc.OperatorsV1alpha1().Subscriptions(subscription.GetNamespace()).Update(context.TODO(), subscription, metav1.UpdateOptions{})
```
[the remark this answers is not in the document]
Comment (anik120) [reply test/e2e/metrics_e2e_test.go]: It doesn't. TIL.

File: test/e2e/metrics_e2e_test.go [hunk 17]
```diff
@@ -117,3 +156,35 @@ func getMetricsFromPod(client operatorclient.ClientInterface, pod *corev1.Pod, p
...
+			Package:                packageName,
+			Channel:                channel,
+			InstallPlanApproval:    approval,
+		},
+		Status: v1alpha1.SubscriptionStatus{
+			InstalledCSV: installed,
+		},
+	}
+
+	subscription, err := crc.OperatorsV1alpha1().Subscriptions(namespace).Create(context.TODO(), subscription, metav1.CreateOptions{})
```
[the remark this answers is not in the document]
Comment (anik120) [reply test/e2e/metrics_e2e_test.go]: Apparently it doesn't. Latest attempt was to update the spec.Package.

File: pkg/controller/operators/catalog/subscription/syncer.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/controller/operators/catalog/subscription/syncer.go:72]: > could you explain what you mean by "from the cache"? 

I mean the one being passed from the informer to this `Sync` function (via `event.Resource()`.

> The only subscription I'm aware of is the one on cluster, which I decided to grab and delete the metrics for before updating the on cluster subscription.

So here you are being given `res.Resource()` which _hopefully_ represents the latest `Subscription` and you are comparing it against the `Subscription` returned from the client, which you _know_ is the latest. Since `res.Resource()` is coming from an informer -- a watch -- chances are pretty good that it's up to date, which means you'll only ever trigger the nested if-statement below in the rare case the `Subscription` is updated between when the event is picked up by the informer and when this sync is called.

File: pkg/metrics/metrics.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (njhale) [reply pkg/metrics/metrics.go:223]: We can potentially fix this in a separate bug.

File: test/e2e/metrics_e2e_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (njhale) [reply test/e2e/metrics_e2e_test.go]: could you actually install an operator? using busybox or hang:10?

[the remark this answers is not in the document]
Comment (anik120) [reply test/e2e/metrics_e2e_test.go]: @njhale do you mean create a subscription for an operator, and then upgrade the operator? 
That could be done, it just feels like an overkill to test metrics being omitted, which is why my original inclination was to keep the test limited to dealing with a standalone subscription object.

[the remark this answers is not in the document]
Comment (njhale) [reply test/e2e/metrics_e2e_test.go]: it does, but this is an e2e test and we are testing behaviors of the highest-scoped API. I'm all in for an alternative if we can think of one (that doesn't involve manually setting the status).

File: pkg/metrics/metrics.go [hunk 18]
```diff
@@ -217,3 +229,34 @@ func EmitCSVMetric(oldCSV *olmv1alpha1.ClusterServiceVersion, newCSV *olmv1alpha
...
```
Comment (ecordell) [inline pkg/metrics/metrics.go:234]: `spec` is not a required field, so this would panic without 

```
	if sub.Spec == nil {
		return
	}
```

File: pkg/metrics/metrics.go [hunk 19]
```diff
@@ -217,3 +229,34 @@ func EmitCSVMetric(oldCSV *olmv1alpha1.ClusterServiceVersion, newCSV *olmv1alpha
...
+		subscriptionSyncCounters[sub.GetName()] = subscriptionSyncLabelValues{
+			installedCSV: sub.Status.InstalledCSV,
+			pkg:          sub.Spec.Package,
+			channel:      sub.Spec.Channel,
+		}
+	}
+}
+
+func DeleteSubsMetric(sub *olmv1alpha1.Subscription) {
+	SubscriptionSyncCount.DeleteLabelValues(sub.GetName(), sub.Status.InstalledCSV, sub.Spec.Channel, sub.Spec.Package)
```
Comment (ecordell) [inline pkg/metrics/metrics.go:245]: The should check for nil on sub.status and sub.spec

File: pkg/metrics/metrics.go [hunk 20]
```diff
@@ -217,3 +229,34 @@ func EmitCSVMetric(oldCSV *olmv1alpha1.ClusterServiceVersion, newCSV *olmv1alpha
...
+}
+
+func UpdateSubsSyncCounterStorage(sub *olmv1alpha1.Subscription) {
+	counterValues := subscriptionSyncCounters[sub.GetName()]
+
```
Comment (ecordell) [inline pkg/metrics/metrics.go:250]: sub.spec and sub.status can be nil
